src/sync/all.rs

```diff
@@ -1388,134 +1473,103 @@ impl<TRq, TSrc, TBl> AllSync<TRq, TSrc, TBl> {
-                            finalized_storage_heap_pages,
-                        }
-                    }
-                    (warp_sync::WarpSync::Finished(_), Some(_)) => unreachable!(),
-                }
+                self.inner = AllSyncInner::GrandpaWarpSync { inner: sync };
+                (user_data.user_data, ResponseOutcome::Queued)
             }
             (
-                AllSyncInner::GrandpaWarpSync {
-                    inner: warp_sync::InProgressWarpSync::StorageGet(sync),
-                },
-                Ok(mut response),
+                AllSyncInner::GrandpaWarpSync { inner: mut sync },
+                Err(_),
+                RequestMapping::WarpSync(request_id),
             ) => {
-                // In this state, we expect the response to be one value. As documented, we panic
-                // if the number of items isn't 1.
-                let value = response.next().unwrap();
-                assert!(response.next().is_none());
-
-                let outcome = sync.inject_value(value.map(iter::once));
-                match outcome {
-                    (warp_sync::WarpSync::InProgress(inner), None) => {
-                        self.inner = AllSyncInner::GrandpaWarpSync { inner };
-                        ResponseOutcome::Queued
-                    }
-                    (warp_sync::WarpSync::InProgress(inner), Some(error)) => {
-                        self.inner = AllSyncInner::GrandpaWarpSync { inner };
-                        ResponseOutcome::WarpSyncError { error }
-                    }
-                    (warp_sync::WarpSync::Finished(success), None) => {
-                        let (
-                            all_forks,
-                            finalized_block_runtime,
-                            finalized_storage_code,
-                            finalized_storage_heap_pages,
-                        ) = self.shared.transition_grandpa_warp_sync_all_forks(success);
-                        self.inner = AllSyncInner::AllForks(all_forks);
-                        ResponseOutcome::WarpSyncFinished {
-                            finalized_block_runtime,
-                            finalized_storage_code,
-                            finalized_storage_heap_pages,
-                        }
-                    }
-                    (warp_sync::WarpSync::Finished(_), Some(_)) => unreachable!(),
-                }
+                let user_data = sync.fail_request(request_id).user_data;
+                // TODO: notify user of the problem
+                self.inner = AllSyncInner::GrandpaWarpSync { inner: sync };
+                (user_data, ResponseOutcome::Queued)
             }
+            // Only the GrandPa warp syncing ever starts GrandPa warp sync requests.
+            (other, _, RequestMapping::Inline(_, _, user_data)) => {
+                self.inner = other;
+                (user_data, ResponseOutcome::Queued) // TODO: no
```

Comment:
what does "no" mean in this context?

Reply:
I think this appears in the diff only because I changed the indentation level, but what I mean is that `Queued` is wrong.
We simply discard the request result and not queue it in any way. `Queued` is kind of used as the default, but ideally it would be something more precise. But because this is just for informative purposes, I haven't fix it properly.